src/client.ts

```diff
@@ -582,17 +607,18 @@ export class Client {
       deliverV2: this.getDeliverV2Callback(),
       consumer_update_query: this.getConsumerUpdateCallback(),
     }
-    return { ...this.params, listeners: connectionListeners, leader: leader, streamName: streamName }
+    return { ...this.params, index: index, listeners: connectionListeners, leader: leader, streamName: streamName }
   }
 
   private async getConnectionOnChosenNode(
     leader: boolean,
+    index: number,
     streamName: string,
     chosenNode: { host: string; port: number },
     metadata: StreamMetadata,
     connectionClosedListener?: ConnectionClosedListener
   ): Promise<Connection> {
-    const connectionParams = this.buildConnectionParams(leader, streamName, connectionClosedListener)
+    const connectionParams = this.buildConnectionParams(leader, streamName, index, connectionClosedListener)
```

Comment:
Maybe we don't need to store the index inside of the connection? It looks like we are just using this index in the caching part, which the connection itself should not be aware of... Could we simply pass it as an arg to cacheConnection?